### PR TITLE
Added a 'logs' command for listing all of the logs. 

### PR DESCRIPTION
**Pull request by jamesemery**
By default it lists failed logs only. Optionally it can expand the logs with GCS or get non-failed logs as well

### CODE REVIEW

File: src/cromshell/logs/command.py [hunk 1]
```diff
@@ -0,0 +1,267 @@
...
+                task_status_font = None
+
+            shardstring = (
+                "" if not sharded else "-shard-" + str(shard_list[i]["shardIndex"])
+            )
+            logs = shard_list[i]["backendLogs"]["log"]
+            if cat_logs:
+                print(
+                    colored(
+                        f"\n\n\n{'='*90}\n{indent}{task}{shardstring}:\t{status}\t {logs}\n{'='*90}",
```
Comment (lbergelson) [inline src/cromshell/logs/command.py:217]: Lets replace this hardcoded 90 with a dynamic padding based on `os.get_terminal_size()`

File: src/cromshell/logs/command.py [hunk 2]
```diff
@@ -0,0 +1,267 @@
...
+                    colored(
+                        f"{indent}{task}{shardstring}:\t{status}\t {logs}",
+                        color=task_status_font,
+                    )
+                )
+            did_print = True
+    return did_print
+
+
+def resolve_and_return_metadata_keys(
```
Comment (lbergelson) [inline src/cromshell/logs/command.py:239]: This isn't used anywhere is it?

File: src/cromshell/logs/command.py [hunk 3]
```diff
@@ -0,0 +1,267 @@
...
+        ],
+        status_params=status_param,
+        dont_expand_subworkflows=dont_expand_subworkflows,
+        cat_logs=print_logs,
+    )
+
+    return 0
+
+
+def obtain_and_print_logs(
```
Comment (lbergelson) [inline src/cromshell/logs/command.py:80]: I would probably reorganize this somewhat there's a function that gets and returns the logs and then a separate one that prints them.

File: tests/unit/test_logs.py [hunk 4]
```diff
@@ -0,0 +1,43 @@
...
+
+        logs_output = logs_command.print_workflow_logs(
+            workflow_metadata=workflow_metadata,
+            expand_sub_workflows=True,
+            indent="",
+            status_keys=status_keys,
+            cat_logs=False,
+        )
+
+        # We should have found a failing output to print
```
Comment (lbergelson) [inline tests/unit/test_logs.py:38]: This comment is out of date I think.

File: src/cromshell/logs/command.py [hunk 5]
```diff
@@ -0,0 +1,267 @@
...
+    did_print = False
+
+    shard_list = workflow_metadata["calls"][task]
+
+    sharded = workflow_metadata["calls"][task][0]["shardIndex"] != -1
+
+    for i in range(len(shard_list)):
+
+        status = shard_list[i]["executionStatus"]
+        if "ALL" in status_keys or status in status_keys:
```
Comment (lbergelson) [inline src/cromshell/logs/command.py:198]: Maybe pull this out as a function get_color_for_status  or something?

File: src/cromshell/logs/command.py [hunk 6]
```diff
@@ -0,0 +1,267 @@
...
+    indent: str,
+    expand_sub_workflows: bool,
+    status_keys: list,
+    cat_logs: bool,
+) -> bool:
+    """
+    Recursively runs through each task of a workflow metadata and calls function to
+    call out to the helper in order to print found logs
+    :param workflow_metadata: Metadata of the workflow to process
+    :param indent: Indent string given as "\t", used to indent print out
```
Comment (lbergelson) [inline src/cromshell/logs/command.py:130]: default isn't actually \t

File: src/cromshell/logs/command.py [hunk 7]
```diff
@@ -0,0 +1,267 @@
...
+        workflow_metadata=workflow_status_json,
+        indent="",
+        expand_sub_workflows=not dont_expand_subworkflows,
+        status_keys=status_params,
+        cat_logs=cat_logs,
+    )
+
+    if not found_logs:
+        print(
+            f"No logs with status {status_params} found for workflow, try adding the argument '-s ALL' to list logs with any status"
```
Comment (lbergelson) [inline src/cromshell/logs/command.py:115]: Woah, can this format a list automatically? neat!

File: src/cromshell/logs/command.py [hunk 8]
```diff
@@ -0,0 +1,267 @@
...
+    """Get a subset of the workflow metadata."""
+
+    LOGGER.info("logs")
+
+    # If no keys were provided then set key_param to empty else
+    # strip trailing comma from keys and split keys by comma
+    status_param = (
+        ["ALL"]
+        if "ALL".lower() in status.lower()
+        else str(status).strip(",").split(",")
```
Comment (lbergelson) [inline src/cromshell/logs/command.py:55]: might be worth being clever and making these case insensitive by manually matching/converting to cromwell values.  Might not be though...

**Review by lbergelson — CHANGES_REQUESTED** [review]
@jamesemery Some minor comments. Looks good otherwise.

File: src/cromshell/logs/command.py [hunk 9]
```diff
@@ -0,0 +1,269 @@
...
+@click.option(
+    "-p",
+    "--print-logs",
+    is_flag=True,
+    default=False,
+    help="Print the contents of the logs to stdout if true. "
+    "Note: This assumes GCS bucket logs with default permissions otherwise this may not work",
+)
+@click.option(
+    "-des",
```
[the remark this answers is not in the document]
Comment (jamesemery) [reply src/cromshell/logs/command.py:32]: Not sure. I'm probably fine with leaving that as a future refactoring to other tools.

File: src/cromshell/logs/command.py [hunk 10]
```diff
@@ -0,0 +1,269 @@
...
+    status_params: list,
+    dont_expand_subworkflows: bool,
+):
+    """Format metadata parameters and obtains metadata from cromwell server"""
+
+    # Combine keys and flags into a dictionary
+    formatted_metadata_parameter = metadata_command.format_metadata_params(
+        list_of_keys=metadata_param,
+        exclude_keys=False,
+        expand_subworkflows=not dont_expand_subworkflows,  # Invert variable
```
[the remark this answers is not in the document]
Comment (jamesemery) [reply src/cromshell/logs/command.py:93]: i don't not understand what you are saying

File: src/cromshell/logs/command.py [hunk 11]
```diff
@@ -0,0 +1,269 @@
...
+
+    # Request workflow metadata
+    workflow_status_json = metadata_command.get_workflow_metadata(
+        meta_params=formatted_metadata_parameter,
+        api_workflow_id=config.cromwell_api_workflow_id,
+        timeout=config.requests_connect_timeout,
+        verify_certs=config.requests_verify_certs,
+    )
+
+    io_utils.pretty_print_json(workflow_status_json, add_color=True)
```
[the remark this answers is not in the document]
Comment (jamesemery) [reply src/cromshell/logs/command.py]: whoops sorry i got rid of that and forgot to push it. That was for debugging

File: src/cromshell/logs/command.py (hunk 2 above)
Comment (jamesemery) [reply src/cromshell/logs/command.py:239]: good point removed. copy past gore

File: src/cromshell/logs/command.py (hunk 7 above)
Comment (jamesemery) [reply src/cromshell/logs/command.py:115]: i know python has some cool things that will occasionally catch fire and fail at runtime in utterly unpredictable ways months later...

**Comment by lbergelson** [comment]
@jamesemery needs rebase please

File: src/cromshell/logs/command.py [hunk 12]
```diff
@@ -180,7 +181,7 @@ def print_task_logs(
     """
     Prints the backend logs from the workflow
     :param task: Name of the task
-    :param indent: Indent string given as "\t", used to indent print out
+    :param indent: Indent string given as a stirng of "\t", used to indent print out
```
Comment (lbergelson) [inline src/cromshell/logs/command.py:184]: typo

**Review by lbergelson — APPROVED** [review]
fix typo and then merge 👍

**Pull request merged**